xml-reader/src/main/java/org/triplea/generic/xml/reader/XmlParser.java

```diff
@@ -0,0 +1,109 @@
+package org.triplea.generic.xml.reader;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.logging.Level;
+import javax.annotation.Nonnull;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamReader;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.java.Log;
+import org.triplea.java.function.ThrowingRunnable;
+
+@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+@Log
+class XmlParser {
+  @Nonnull private final String tagName;
+
+  public static XmlParser tag(final String tagName) {
+    return new XmlParser(tagName.toUpperCase());
+  }
+
+  public TagParser childTagHandler(final String childTagName, final ThrowingRunnable handler) {
+    return new TagParser(tagName, new HashMap<>()).childTagHandler(childTagName, handler);
+  }
+
+  public BodyParser bodyHandler(final Consumer<String> bodyHandler) {
+    return new BodyParser(tagName, bodyHandler);
+  }
+
+  @AllArgsConstructor(access = AccessLevel.PRIVATE)
+  public static class TagParser {
+    private final String currentTag;
+    private final Map<String, ThrowingRunnable> childTagHandlers;
+
+    public TagParser(String currentTag) {
+      this.currentTag = currentTag.toUpperCase();
+      this.childTagHandlers = new HashMap<>();
+    }
+
+    public TagParser childTagHandler(final String childTagName, final ThrowingRunnable tagHandler) {
+      childTagHandlers.put(childTagName.toUpperCase(), tagHandler);
+      return this;
+    }
+
+    public void parse(final XMLStreamReader xmlReader) throws XMLStreamException {
+      new GenericParser(currentTag, childTagHandlers, body -> {}).parse(xmlReader);
+    }
+  }
+
+  @AllArgsConstructor(access = AccessLevel.PRIVATE)
+  public static class BodyParser {
+    private final String currentTag;
+    private Consumer<String> bodyHandler;
+
+    public void parse(final XMLStreamReader xmlReader) throws XMLStreamException {
+      new GenericParser(currentTag, Map.of(), bodyHandler).parse(xmlReader);
+    }
+  }
+
+  @Log
+  @AllArgsConstructor(access = AccessLevel.PRIVATE)
+  public static class GenericParser {
+    private final String currentTag;
+    private final Map<String, ThrowingRunnable> childTagHandlers;
+    private Consumer<String> bodyHandler;
+
+    public void parse(final XMLStreamReader streamReader) throws XMLStreamException {
+      if (log.isLoggable(Level.FINE)) {
+        log.fine("Processing tag: " + currentTag);
+      }
```

Comment:
Does this if statement make a difference?
AFAIK log.fine is no-op when this level is not loggable

Reply:
The if statement is to avoid evaluation of arguments.

Say our logging method is:
```
void fine(String statementToLog) {
   if(log.isLoggable(this.currentLogLevel()) {
          // do some logging
   }
}
```

If you invoke that with:
```
log.fine("debug statement: " + expensiveToString.toString());
```

Then in the above you'll evaluate: `"debug statement: " + expensiveToString.toString()` to then pass that to the logging statement which then on-ops. So by wrapping the statement with an if you are avoiding the parameter evaluation. When a method is called, first the parameters are evaluated and then those are passed to the method.

Since we have the concatenation of two strings, we are not saving much at all. Our default log level is only info, so this is further not super useful to have..

Reply:
I believe you can use the overload [`Logger#fine(Supplier<String>)`](https://docs.oracle.com/en/java/javase/11/docs/api/java.logging/java/util/logging/Logger.html#fine(java.util.function.Supplier)) to avoid building the message when the current log level is above `FINE`.

Reply:
Good to know that's been built-in now @ssoloff, having to wrap log statements with if conditions always seems kinda gross.